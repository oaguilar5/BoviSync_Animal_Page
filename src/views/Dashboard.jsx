/*!

=========================================================
* Animal Page v1.0.0
* Dashboard View
=========================================================

* Product Page: https://github.com/oaguilar5/BoviSync_Animal_Page
* Free License

* Author: Oscar Aguilar
* Created: 11/25/2020

=========================================================


*/


import React from "react";
import Section from "../components/Section.jsx"
// reactstrap components
import { 
    Button,
    Card, 
    CardHeader, 
    CardBody,
    Col,
    Input,
    Label,
    Row,
    Modal,
    ModalHeader,
    ModalBody,
    ModalFooter,
    Navbar,
    NavbarToggler,
    Nav,
    NavItem,
    NavLink,
    Collapse,
    UncontrolledDropdown,
    DropdownItem,
    DropdownToggle,
    DropdownMenu
} from "reactstrap";


var className = "dashboard-view"
var colors = [
    "auto", 
    "lightsalmon", 
    "lightseagreen", 
    "lightgoldenrodyellow", 
    "lightslategray", 
    "lightcoral", 
    "brown", 
    "mediumaquamarine", 
    "goldenrod", 
    "mediumorchid"
]


class Dashboard extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            welcomeMsg: "Welcome! Please specify your preferences below",
            infoMsg: "Hover over any table header to view the data item name. Hover over any table data item to view its description.",
            infoModal: false,
            modalMsg: "",
            itemMeta: {},
            pageMeta: [],
            animalData: {},
            decimals: "",
            thresholdHigh: "",
            colorHigh: "",
            thresholdLow: "",
            colorLow: "",
            dataUnknown: "Data Unknown",
            headerColors: colors,
            noData: ""
        };
      }

    /**
     * Name: componentDidMount
     * Description: React component lifecycle funciton called when component renders
     * 
     * @params none
     * @returns none
    **/
    componentDidMount() {
        this.fetchAppData();
    }
    
    /**
     * Name: fetchAppData
     * Description: retrieves meta data and report item JSON files to determine how to render the page
     * 
     * @params none
     * @requires JSON file links provided by BoviSync
     * @returns none - sets the state variables of the component
    **/
    fetchAppData = () => {
        //make async calls to retrieve each item and set its state

        //item meta
        fetch(this.props.itemMetaUrl)
            .then(response => response.json())
            .then(data => this.itemMetaReconstruct(data))
            .catch(err => {
                //display an error message to the user
                this.setState({infoModal: true, modalMsg: "Error retrieving Item Meta Data, please try again"})
                //DEBUG
                console.log("Dashboard.jsx: error retrieving item meta. Error: " + err)
            })

        //page meta
        fetch(this.props.pageMetaUrl)
            .then(response => response.json())
            .then(data => {
                //check to ensure the 'sections' array is present
                if (data.sections) {
                    this.setState({pageMeta: data.sections})
                } else {
                    //present error message
                    this.setState({infoModal: true, modalMsg: "Error: no sections present in Page Meta Data", noData: "No Sections Available"})
                }
            })
            .catch(err => {
                //display an error message to the user
                this.setState({infoModal: true, modalMsg: "Error retrieving Page Meta Data, please try again", noData: "No Sections Available"})
                //DEBUG
                console.log("Dashboard.jsx: error retrieving page meta. Error: " + err)
            })

        //animal data
        fetch(this.props.animalDataUrl)
            .then(response => response.json())
            .then(data => this.setState({animalData: data}))
            .catch(err => {
                //display an error message to the user
                this.setState({infoModal: true, modalMsg: "Error retrieving Animal Data, please try again"})
                //DEBUG
                console.log("Dashboard.jsx: error retrieving animal data. Error: " + err)
            })
    }

    /**
     * Name: itemMetaReconstruct
     * Description: reconstructs itemMeta as a Map where each key is the shortName
     * 
     * @params itemMeta - object retrieved from json fetch
     * @requires JSON file item meta to be an array of objects which should contain shortName
     * @returns none - sets the itemMeta state variable
    **/
   itemMetaReconstruct = itemMeta => {
        try {
            let itemMetaMap = {};
            for (let i=0;i<itemMeta.length;i++) {
                let thisItem = {[itemMeta[i].shortName]: itemMeta[i]}
                Object.assign(itemMetaMap, thisItem)
            }
            //set the state variable
            this.setState({itemMeta: itemMetaMap})
        } catch (err) {
            //present an error
            console.log("Dashboard.jsx: encountered error in itemMetaReconstruct(). Error: " + err)
        }
   }

   /**
     * Name: toggleModal
     * Description: negates the value of the modal to open/close
     * 
     * @params none
     * @requires none
     * @returns none - sets the state variable
    **/
   toggleModal = () => {
        this.setState(prevState => { return { infoModal: !prevState.infoModal } })
    }


    /**
     * Name: inputOnBlur
     * Description: Special method assigned to the two threshold input elements...
     * Used to evaluate the threshold values to ensure low is not greater than high and vice-versa
     * 
     * @params evt - the event triggered by focusing on the element
     * @requires none
     * @returns none - sets the state variable
    **/
    inputOnBlur = evt => {
        try {
            let id = evt.target.id;
            let value = evt.target.value;
            value = parseInt(value);
            //check if a threshold value was updated that it doesn't breach the other's limit
            let thresholdLow = this.state.thresholdLow;
            let thresholdHigh = this.state.thresholdHigh;
            //if the values were breached, change the threshold value
            if (thresholdHigh && id === "thresholdLow" && value > thresholdHigh) {
                this.setState({thresholdHigh: value})
            } else if (thresholdLow && id === "thresholdHigh" && value < thresholdLow) {
                this.setState({thresholdLow: value})
            }
        } catch (err) {
            console.log("Dashboard.jsx: encountered error in inputOnBlur(). Error: " + err)
        }
    }

    /**
     * Name: inputOnChange
     * Description: Basic method assigned to all inputs to change their value state variable
     * 
     * @params none
     * @requires none
     * @returns none - sets the state variable
    **/
   inputOnChange = evt => {
       try {
            let id = evt.target.id;
            let value = evt.target.value;
            let type = evt.target.type;
            if (type === "select") {
                    value = evt.target.options[evt.target.selectedIndex].value;
            } else if (type === "number") {
                //sanitize numbers
                value = parseInt(value);
            }
            this.setState({[id]: value})
       } catch (err) {
        console.log("Dashboard.jsx: encountered error in inputOnChange(). Error: " + err)
       }
        
   }

  render() {
    return (
      <>
        <div className={className}>
        <Navbar color="light" light expand="lg">
          <div className="logo">
            <a href="/app"><img src="/images/logo_1.png" alt="Animal Page" /></a>
          </div>
            <NavbarToggler />
            <Collapse navbar>
              <Nav className="mr-auto" navbar>
                <NavItem>
                  <NavLink href="/app">Home</NavLink>
                </NavItem>
                <NavItem>
                  <NavLink href="/app">My Farm</NavLink>
                </NavItem>
                <NavItem>
                  <NavLink href="/app">Dairy Central</NavLink>
                </NavItem>
                <NavItem>
                  <NavLink href="/app">Search</NavLink>
                </NavItem>

              </Nav>
              <div className="account-link">
                <UncontrolledDropdown>
                  <DropdownToggle
                    color="default"
                    data-toggle="dropdown"
                    caret
                  >
                    <div className="user-photo">
                      <img alt="User" src="/images/user-icon.png" />
                    </div>
                    <p className="d-lg-none">Log out</p>
                  </DropdownToggle>
                  <DropdownMenu className="dropdown-navbar" right tag="ul">
                    <DropdownItem header >John Doe</DropdownItem>
                    <DropdownItem divider tag="li" />
                    <NavLink tag="li" >
                      <DropdownItem className="nav-item" href="/app">Log out</DropdownItem>
                    </NavLink>
                  </DropdownMenu>
                </UncontrolledDropdown>
              </div>
            </Collapse>
          </Navbar>
            <Modal isOpen={this.state.infoModal} toggle={this.toggleModal}>
                <ModalHeader toggle={this.toggleModal}>Notice</ModalHeader>
                <ModalBody>
                    {this.state.modalMsg}
                </ModalBody>
                <ModalFooter>
                    <Button color="primary" onClick={this.toggleModal}>Ok</Button>{' '}
                </ModalFooter>
            </Modal>
            <div className="preferences">
                <Card>
                    <CardHeader><b>Preferences</b></CardHeader>
                    <CardBody>
                        <p>{this.state.welcomeMsg}</p>
                        <p>{this.state.infoMsg}</p>
                        <Row>
                            <Col md="2">
                                <Label>Decimal Places</Label>
                                <Input 
                                    type="number" 
                                    id="decimals" 
                                    min="0" 
                                    max="10" 
                                    value={this.state.decimals} 
                                    onChange={this.inputOnChange}
                                />
                            </Col>
                            <Col md="2">
                                <Label>Mark values over</Label>
                                <Input 
                                    type="number" 
                                    id="thresholdHigh" 
                                    value={this.state.thresholdHigh} 
                                    onChange={this.inputOnChange}
                                    onBlur={this.inputOnBlur}
                                />
                            </Col>
                            <Col md="2">
                                <Label>With Color</Label>
                                <Input 
                                    type="select" 
                                    id="colorHigh" 
                                    value={this.state.colorHigh} 
                                    onChange={this.inputOnChange} 
                                    
                                >
                                    <option></option>
                                    <option>Green</option>
                                    <option>Red</option>
                                    <option>Blue</option>
                                </Input>
                            </Col>
                            <Col md="2">
                                <Label>Mark values under</Label>
                                <Input 
                                    type="number" 
                                    id="thresholdLow" 
                                    value={this.state.thresholdLow} 
                                    onChange={this.inputOnChange}
                                    onBlur={this.inputOnBlur}
                                />
                            </Col>
                            <Col md="2">
                                <Label>With Color</Label>
                                <Input 
                                    type="select" 
                                    id="colorLow" 
                                    value={this.state.colorLow} 
                                    onChange={this.inputOnChange} 
                                    onBlur={this.inputOnBlur}
                                >
                                    <option></option>
                                    <option>Green</option>
                                    <option>Red</option>
                                    <option>Blue</option>
                                </Input>
                            </Col>
                            <Col md="2">
                                <Label>Specify Unknown Values As</Label>
                                <Input 
                                    type="text" 
                                    id="dataUnknown" 
                                    value={this.state.dataUnknown} 
                                    onChange={this.inputOnChange}
                                />
                            </Col>
                        </Row>
                    </CardBody>
                </Card>
            </div>
            {this.state.pageMeta.length > 0 &&
                [...this.state.pageMeta].map((item, index) => (
                    <Section
                        key={Date.now() + "_" + item.label}
                        headerName={item.label}
                        items={item.items}
                        itemMeta={this.state.itemMeta}
                        animalData={this.state.animalData}
                        decimals={this.state.decimals}
                        thresholdHigh={this.state.thresholdHigh}
                        thresholdLow={this.state.thresholdLow}
                        colorHigh={this.state.colorHigh}
                        colorLow={this.state.colorLow}
                        dataUnknown={this.state.dataUnknown}
                        headerColor={this.state.headerColors[index % 10]}
                    />
                ))
            }
            {this.state.pageMeta.length === 0 &&
                <h1 className="text-muted no-data">{this.state.noData}</h1>
            }
        </div>
      </>
    );
  }
}

export default Dashboard;