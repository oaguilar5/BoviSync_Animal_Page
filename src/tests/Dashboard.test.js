import React from 'react';
import { shallow } from 'enzyme';
import { render, screen } from '@testing-library/react';
import Dashboard from '../views/Dashboard.jsx';
import fetchMock from "jest-fetch-mock";

var itemMetaUrl = 'https://bovisync.bitbucket.io/sample_data/item_meta.json'
    var pageMetaUrl = 'https://bovisync.bitbucket.io/sample_data/page_meta.json'
    var animalDataUrl = 'https://bovisync.bitbucket.io/sample_data/animal_data.json'

fetchMock.enableMocks();
beforeEach(() => {
    fetch.resetMocks();
  });



test('should test Dashboard a count', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    expect(wrapper.find("a").length).toEqual(1);
});

test('should test Dashboard img count', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    expect(wrapper.find("img").length).toEqual(2);
});

test('should test Dashboard p count', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    expect(wrapper.find("p").length).toEqual(3);
});

test('should test Dashboard label count', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    expect(wrapper.find("Label").length).toEqual(6);
});

test('should test Dashboard input count', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    expect(wrapper.find("Input").length).toEqual(6);
});

test('should test Dashboard card count', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    expect(wrapper.find("Card").length).toEqual(1);
});

test('should test Dashboard paragraph 1', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    render(<Dashboard />);
    const pElement = screen.getByText(/Welcome! Please specify your preferences below/i);
    expect(pElement).toBeInTheDocument();
});

test('should test Dashboard paragraph 2', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    render(<Dashboard />);
    const pElement = screen.getByText(/Hover over any table header to view the data item name. Hover over any table data item to view its description./i);
    expect(pElement).toBeInTheDocument();
});

test('should test Dashboard component snapshot', () => {
    
    fetch.mockReject(() => Promise.reject("API is down"));
 const wrapper = shallow(<Dashboard itemMetaUrl={itemMetaUrl} pageMetaUrl={pageMetaUrl} animalDataUrl={animalDataUrl} />);
 expect(wrapper).toMatchSnapshot();
 
});


test('should test Dashboard called fetch 3 times', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    render(<Dashboard />);
    expect(fetch).toHaveBeenCalledTimes(3);
});

test('should test Dashboard called fetch with item meta', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    render(<Dashboard itemMetaUrl={itemMetaUrl}/>);
    expect(fetch).toHaveBeenCalledWith(itemMetaUrl);
});

test('should test Dashboard called fetch with page meta', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    render(<Dashboard pageMetaUrl={pageMetaUrl} />);
    expect(fetch).toHaveBeenCalledWith(pageMetaUrl);
});

test('should test Dashboard called fetch with animal data', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    render(<Dashboard animalDataUrl={animalDataUrl} />);
    expect(fetch).toHaveBeenCalledWith(animalDataUrl);
});

test('should test Dashboard item meta length', () => {
    let itemMeta = [
        {
            "dataType": "Integer",
            "name": "Eartag",
            "tags": [],
            "aggregationFunction": "count",
            "shortName": "EART",
            "link": "/cowpage/id/ANIMALID/",
            "units": null,
            "description": "The animal's eartag number. Double clicking this cell will open the animal page."
          },
          {
            "dataType": "Integer",
            "name": "Pen",
            "tags": [],
            "aggregationFunction": null,
            "shortName": "Pen",
            "link": "",
            "units": null,
            "description": "The pen which the animal is in"
          },
    ]
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'itemMetaReconstruct')
    instance.itemMetaReconstruct(itemMeta);
    expect(Object.keys(instance.state.itemMeta).length).toEqual(2);
});


test('should test Dashboard input decimals onChange', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'inputOnChange')
    wrapper.find('#decimals').simulate('change', { target: { id: 'decimals', value: 5 } });
    expect(instance.state.decimals).toEqual(5);
});

test('should test Dashboard input decimals onChange', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'inputOnChange')
    wrapper.find('#decimals').simulate('change', { target: { id: 'decimals', value: 5 } });
    expect(instance.state.decimals).toEqual(5);
});

test('should test Dashboard input thresholdHigh onChange', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'inputOnChange')
    wrapper.find('#thresholdHigh').simulate('change', { target: { id: 'thresholdHigh', value: 5 } });
    expect(instance.state.thresholdHigh).toEqual(5);
});

test('should test Dashboard input thresholdLow onChange', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'inputOnChange')
    wrapper.find('#thresholdLow').simulate('change', { target: { id: 'thresholdLow', value: 5 } });
    expect(instance.state.thresholdLow).toEqual(5);
});

test('should test Dashboard input thresholdHigh onBlur after breaching the bounds', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'inputOnBlur')
    wrapper.find('#thresholdLow').simulate('change', { target: { id: 'thresholdLow', value: 10 } });
    wrapper.find('#thresholdHigh').simulate('blur', { target: { id: 'thresholdHigh', value: 5 } });
    expect(instance.state.thresholdLow).toEqual(5);
});

test('should test Dashboard input thresholdLow onBlur after breaching the bounds', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'inputOnBlur')
    wrapper.find('#thresholdHigh').simulate('change', { target: { id: 'thresholdHigh', value: 5 } });
    wrapper.find('#thresholdLow').simulate('blur', { target: { id: 'thresholdLow', value: 10 } });
    expect(instance.state.thresholdHigh).toEqual(10);
});

test('should test Dashboard input colorHigh onChange', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'inputOnChange')
    wrapper.find('#colorHigh').simulate('change', { target: { id: 'colorHigh', value: "Red" } });
    expect(instance.state.colorHigh).toEqual("Red");
});

test('should test Dashboard input colorLow onChange', () => {
    fetch.mockReject(() => Promise.reject("API is down"));
    const wrapper = shallow(<Dashboard />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'inputOnChange')
    wrapper.find('#colorLow').simulate('change', { target: { id: 'colorLow', value: "Green" } });
    expect(instance.state.colorLow).toEqual("Green");
});




