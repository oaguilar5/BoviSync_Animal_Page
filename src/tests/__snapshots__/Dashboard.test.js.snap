// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test Dashboard component snapshot 1`] = `
<Fragment>
  <div
    className="dashboard-view"
  >
    <Navbar
      color="light"
      expand="lg"
      light={true}
      tag="nav"
    >
      <div
        className="logo"
      >
        <a
          href="/app"
        >
          <img
            alt="Animal Page"
            src="/images/logo_1.png"
          />
        </a>
      </div>
      <NavbarToggler
        tag="button"
        type="button"
      />
      <Collapse
        appear={false}
        enter={true}
        exit={true}
        in={false}
        isOpen={false}
        mountOnEnter={false}
        navbar={true}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        tag="div"
        timeout={350}
        unmountOnExit={false}
      >
        <Nav
          className="mr-auto"
          navbar={true}
          tag="ul"
          vertical={false}
        >
          <NavItem
            tag="li"
          >
            <NavLink
              href="/app"
              tag="a"
            >
              Home
            </NavLink>
          </NavItem>
          <NavItem
            tag="li"
          >
            <NavLink
              href="/app"
              tag="a"
            >
              My Farm
            </NavLink>
          </NavItem>
          <NavItem
            tag="li"
          >
            <NavLink
              href="/app"
              tag="a"
            >
              Dairy Central
            </NavLink>
          </NavItem>
          <NavItem
            tag="li"
          >
            <NavLink
              href="/app"
              tag="a"
            >
              Search
            </NavLink>
          </NavItem>
        </Nav>
        <div
          className="account-link"
        >
          <UncontrolledDropdown>
            <DropdownToggle
              aria-haspopup={true}
              caret={true}
              color="default"
              data-toggle="dropdown"
            >
              <div
                className="user-photo"
              >
                <img
                  alt="User"
                  src="/images/user-icon.png"
                />
              </div>
              <p
                className="d-lg-none"
              >
                Log out
              </p>
            </DropdownToggle>
            <DropdownMenu
              className="dropdown-navbar"
              flip={true}
              right={true}
              tag="ul"
            >
              <DropdownItem
                header={true}
                tag="button"
                toggle={true}
              >
                John Doe
              </DropdownItem>
              <DropdownItem
                divider={true}
                tag="li"
                toggle={true}
              />
              <NavLink
                tag="li"
              >
                <DropdownItem
                  className="nav-item"
                  href="/app"
                  tag="button"
                  toggle={true}
                >
                  Log out
                </DropdownItem>
              </NavLink>
            </DropdownMenu>
          </UncontrolledDropdown>
        </div>
      </Collapse>
    </Navbar>
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransition={
        Object {
          "mountOnEnter": true,
          "timeout": 150,
        }
      }
      centered={false}
      container="body"
      fade={true}
      isOpen={false}
      keyboard={true}
      modalTransition={
        Object {
          "timeout": 300,
        }
      }
      onClosed={[Function]}
      onOpened={[Function]}
      returnFocusAfterClose={true}
      role="dialog"
      scrollable={false}
      toggle={[Function]}
      unmountOnClose={true}
      zIndex={1050}
    >
      <ModalHeader
        charCode={215}
        closeAriaLabel="Close"
        tag="h5"
        toggle={[Function]}
        wrapTag="div"
      >
        Notice
      </ModalHeader>
      <ModalBody
        tag="div"
      />
      <ModalFooter
        tag="div"
      >
        <Button
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Ok
        </Button>
         
      </ModalFooter>
    </Modal>
    <div
      className="preferences"
    >
      <Card
        tag="div"
      >
        <CardHeader
          tag="div"
        >
          <b>
            Preferences
          </b>
        </CardHeader>
        <CardBody
          tag="div"
        >
          <p>
            Welcome! Please specify your preferences below
          </p>
          <p>
            Hover over any table header to view the data item name. Hover over any table data item to view its description.
          </p>
          <Row
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <Col
              md="2"
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Label
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                Decimal Places
              </Label>
              <Input
                id="decimals"
                max="10"
                min="0"
                onChange={[Function]}
                type="number"
                value=""
              />
            </Col>
            <Col
              md="2"
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Label
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                Mark values over
              </Label>
              <Input
                id="thresholdHigh"
                onBlur={[Function]}
                onChange={[Function]}
                type="number"
                value=""
              />
            </Col>
            <Col
              md="2"
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Label
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                With Color
              </Label>
              <Input
                id="colorHigh"
                onChange={[Function]}
                type="select"
                value=""
              >
                <option />
                <option>
                  Green
                </option>
                <option>
                  Red
                </option>
                <option>
                  Blue
                </option>
              </Input>
            </Col>
            <Col
              md="2"
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Label
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                Mark values under
              </Label>
              <Input
                id="thresholdLow"
                onBlur={[Function]}
                onChange={[Function]}
                type="number"
                value=""
              />
            </Col>
            <Col
              md="2"
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Label
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                With Color
              </Label>
              <Input
                id="colorLow"
                onBlur={[Function]}
                onChange={[Function]}
                type="select"
                value=""
              >
                <option />
                <option>
                  Green
                </option>
                <option>
                  Red
                </option>
                <option>
                  Blue
                </option>
              </Input>
            </Col>
            <Col
              md="2"
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Label
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                Specify Unknown Values As
              </Label>
              <Input
                id="dataUnknown"
                onChange={[Function]}
                type="text"
                value="Data Unknown"
              />
            </Col>
          </Row>
        </CardBody>
      </Card>
    </div>
    <h1
      className="text-muted no-data"
    />
  </div>
</Fragment>
`;
